### ISSUE TITLE
[Easy] Add tests throughout the engine.

### ISSUE THREAD

**Issue by jojolepro** [amethyst/amethyst#819]
Task:
Add tests for systems and various functions throughout the engine. You can do a single test like you can do dozens, that's up to you.

Reason:
- Prevent runtime bugs as much as possible
- Observe where usage of our systems are problematic (too complex, too many dependencies, unresolvable errors, not enough configurability, etc...)

We almost have no test in amethyst, but here is the best example:
https://github.com/amethyst/amethyst/blob/85a68a5fe17779fc4bc85077f64531e46fbc17de/amethyst_assets/src/prefab/mod.rs#L451

**Comment by Xaeroxe** [comment]
Please note that any operation requiring a GPU or a window will not be testable on CI at this time.  Please set these tests to auto ignore when you write them.

**Comment by azriel91** [comment]
yo! Saw this issue and thought you guys might want [this amethest beast](https://github.com/azriel91/amethyst_test_support).

TL;DR: it lets you write tests like this:

```rust
// Scroll down for types
#[test]
fn system_increases_component_value_by_one() {
    let effect_fn = |world: &mut World| {
        let entity = world.create_entity().with(MyComponent(0)).build();

        world.add_resource(EffectReturn(entity));
    };
    let assertion = |world: &mut World| {
        let entity = world.read_resource::<EffectReturn<Entity>>().0.clone();

        let my_component_storage = world.read_storage::<MyComponent>();
        let my_component = my_component_storage
            .get(entity)
            .expect("Entity should have a `MyComponent` component.");

        // If the system ran, the value in the `MyComponent` should be 1.
        assert_eq!(1, my_component.0);
    };

    assert!(
        AmethystApplication::blank()
            .with_system(MySystem, "my_system", &[])
            .with_effect(effect_fn)
            .with_assertion(assertion)
            .run()
            .is_ok()
    );
}

struct MyComponent(pub i32);
impl Component for MyComponent {
    type Storage = DenseVecStorage<Self>;
}

#[derive(Debug)]
struct MySystem;
type MySystemData<'s> = WriteStorage<'s, MyComponent>;
impl<'s> System<'s> for MySystem {
    type SystemData = MySystemData<'s>;
    fn run(&mut self, mut my_component_storage: Self::SystemData) {
        for mut my_component in (&mut my_component_storage).join() {
            my_component.0 += 1
        }
    }
}
```

Features:

* Lets you test `Bundle`s, `State`s, `System`s, types that add/mutate things in `World` with very little boilerplate.
* The test harness itself is extensively tested.

---

It evolves pretty quickly in my game repo with no guarantees of stability, that's why I never pushed it anywhere, but it lets you test types without having to set up the application and state boilerplate. If you want to take it, make it good<sup>TM</sup>, and put it in this repo, by all means. It does contain Rust generics black magic to make it more ergonomic for the user (user base size = 1, me, haha).

I probably will keep evolving my own copy in my game repo since I'm impatient about pushing code, and maybe if/when it stabilizes I can let you guys know.

**Comment by jojolepro** [comment]
This looks very good! Definitely something that could be worth adding to the engine. Maybe it could even be used to create full test games and examples that are easier to access for newcomers. Imagine making a game with 
```
fn main() -> Result<()>{
AmethystApplication::render_base("Game 1", true).with_system(TestSystem,"test_system",&[]).run()?
}
```

**Comment by torkleyy** [comment]
We can test graphics on CI, using a headless context. I'm fairly certain this works with at least Mesa.

**Comment by torkleyy** [comment]
I've implemented it for the gfx repo months ago, not sure if they're still using it though.

**Comment by fhaynes** [comment]
Closing this as it seems to have served its purpose. Moving forward, we'll have to focus on improving testing in specific areas.
